File: pmstrat/pmstrat/strategies/momentum_fade.py
"""Momentum-Fade Spike Detector.

Watches a small set of tokens for volume spikes. When the last `SHORT_WINDOW`
seconds of volume on a token exceeds `SPIKE_MULTIPLIER` × the recent baseline
volume (computed from the prior `LONG_WINDOW` seconds), raise a `Signal::Alert`
proposing a fade trade: a Sell at the current mid. The human reviews the
alert (typically by checking whether @Polymarket / Trump / etc. just tweeted
about the market) and approves or rejects via `pmt engine approve|reject`.

This strategy doesn't *itself* identify whether the move is informed or
sheep-driven — that's the human-touch step. It just flags candidates where
the volume profile looks like a coordinated pump worth a closer look.

Designed for the engine's Phase 5 alert pipeline: the strategy never places
a real order — every action is gated on human approval.
"""

from decimal import Decimal

from ..dsl import strategy
from ..signal import Alert, Hold, Sell, Signal, Urgency


# Tokens to watch. Start small with high-volume markets where pumps actually
# happen. Polymarket's spike-prone markets tend to be politics, crypto, and
# event-driven (election outcomes, court rulings, AI announcements, etc.).
#
# Hardcoded for now — Phase 6 ships an MVP that exercises the pipeline.
# Future revisions could add Signal::Subscribe on startup based on a
# scanner query against gamma.
WATCH_TOKENS: list[str] = [
    # Hantavirus pandemic NO — we have data for this already
    "95212449865986159112377413335252801281670333750637442556685159781445406848396",
]

# Spike detection knobs.
SHORT_WINDOW = 60     # seconds — the window we evaluate for "recent" volume
LONG_WINDOW = 900     # seconds — baseline window (15 min)
SPIKE_MULTIPLIER = Decimal("5.0")   # short-window rate / baseline rate threshold
MIN_SHORT_VOLUME = Decimal("100")   # ignore tiny markets — at least N shares
                                    # in the short window before we bother

ALERT_TTL_SECS = 600   # 10-minute window for the human to act
SUGGESTED_SIZE = Decimal("100")     # what to fade with — modest


@strategy(
    name="momentum_fade",
    tokens=WATCH_TOKENS,
    tick_interval_ms=5000,   # 5s — fast enough to catch a 60s spike,
                              # slow enough to not hammer
    params={
        "SHORT_WINDOW": SHORT_WINDOW,
        "LONG_WINDOW": LONG_WINDOW,
        "SPIKE_MULTIPLIER": SPIKE_MULTIPLIER,
        "MIN_SHORT_VOLUME": MIN_SHORT_VOLUME,
        "SUGGESTED_SIZE": SUGGESTED_SIZE,
        "ALERT_TTL_SECS": ALERT_TTL_SECS,
    },
)
def on_tick(ctx) -> list[Signal]:
    signals: list[Signal] = []

    for token_id in WATCH_TOKENS:
        book = ctx.book(token_id)
        if book is None:
            continue
        best_bid = book.best_bid
        if best_bid is None:
            continue
        best_ask = book.best_ask
        if best_ask is None:
            continue

        short_vol = ctx.volume_in_window(token_id, SHORT_WINDOW)
        if short_vol < MIN_SHORT_VOLUME:
            continue

        long_vol = ctx.volume_in_window(token_id, LONG_WINDOW)
        # Baseline = long-window volume excluding the short window, normalized
        # to a per-second rate so it's comparable to the short-window rate.
        baseline_vol = long_vol - short_vol
        if baseline_vol <= Decimal("0"):
            # No baseline = first observation; can't say it's a spike.
            continue
        baseline_secs = LONG_WINDOW - SHORT_WINDOW
        baseline_rate = baseline_vol / Decimal(baseline_secs)
        short_rate = short_vol / Decimal(SHORT_WINDOW)
        if short_rate < baseline_rate * SPIKE_MULTIPLIER:
            continue

        # Spike detected. Propose a fade.
        mid = (best_bid + best_ask) / Decimal("2")
        # Round suggested sell price down to a tick we know is safe; the
        # engine's per-market tick rounding will adjust if needed.
        suggested_price = mid - Decimal("0.005")
        if suggested_price <= Decimal("0.01"):
            suggested_price = Decimal("0.01")

        # Dedupe key bucketed to the minute so we re-alert if the spike
        # persists into a new minute, but not every tick.
        bucket = int(ctx.timestamp.timestamp()) // 60
        signals.append(
            Alert(
                reason=f"vol spike: short {short_vol} / baseline {baseline_vol:.0f} ({float(short_rate/baseline_rate):.1f}x)",
                suggested=Sell(
                    token_id=token_id,
                    size=SUGGESTED_SIZE,
                    price=suggested_price,
                    urgency=Urgency.MEDIUM,
                ),
                ttl_secs=ALERT_TTL_SECS,
                dedupe_key=f"momentum_fade-{token_id[:16]}-{bucket}",
            )
        )

    if not signals:
        return [Hold()]
    return signals

File: pmengine/src/strategies/momentum_fade.rs
//! Momentum-Fade Spike Detector (hand-written; pmstrat source is the spec).
//!
//! Mirrors `pmstrat/strategies/momentum_fade.py`. Written by hand because the
//! transpiler doesn't yet handle a few patterns the Python uses (f-strings,
//! Option-unwrapping flow, Decimal-from-int coercion). Treat the Python file
//! as the source of truth for intent; this is the executable Rust copy.
//!
//! On each tick, for every watched token:
//!   - Compute volume in the last SHORT_WINDOW seconds (from MarketDataHub
//!     rolling buffer, exposed via ctx.volume_in_window).
//!   - Skip if volume is below MIN_SHORT_VOLUME (filter dust).
//!   - Compute baseline rate over LONG_WINDOW − SHORT_WINDOW prior seconds.
//!   - If short-window rate > baseline_rate × SPIKE_MULTIPLIER, emit a
//!     Signal::Alert proposing a Sell at mid − 0.005 (fade the move).
//!   - Dedupe by token + minute bucket so the same spike doesn't re-alert
//!     every 5s tick.

use crate::position::Fill;
use crate::strategy::{Signal, Strategy, StrategyContext, Urgency};
use rust_decimal::Decimal;
use rust_decimal_macros::dec;

const SHORT_WINDOW: i64 = 60;
const LONG_WINDOW: i64 = 900;
const SPIKE_MULTIPLIER: Decimal = dec!(5);
const MIN_SHORT_VOLUME: Decimal = dec!(100);
const SUGGESTED_SIZE: Decimal = dec!(100);
const ALERT_TTL_SECS: i64 = 600;

const HANTA_NO_TOKEN: &str =
    "95212449865986159112377413335252801281670333750637442556685159781445406848396";

pub struct MomentumFade {
    id: String,
    tokens: Vec<String>,
}

impl MomentumFade {
    pub fn new() -> Self {
        Self {
            id: "momentum_fade".to_string(),
            tokens: vec![HANTA_NO_TOKEN.to_string()],
        }
    }
}

impl Default for MomentumFade {
    fn default() -> Self {
        Self::new()
    }
}

impl Strategy for MomentumFade {
    fn id(&self) -> &str {
        &self.id
    }

    fn subscriptions(&self) -> Vec<String> {
        self.tokens.clone()
    }

    fn tick_interval_ms(&self) -> u64 {
        5000
    }

    fn on_tick(&mut self, ctx: &StrategyContext) -> Vec<Signal> {
        let mut signals: Vec<Signal> = Vec::new();
        for token_id in &self.tokens {
            let book = match ctx.order_books.get(token_id) {
                Some(b) => b,
                None => continue,
            };
            let (Some(best_bid), Some(best_ask)) = (book.best_bid(), book.best_ask()) else {
                continue;
            };

            let short_vol = ctx.volume_in_window(token_id, SHORT_WINDOW);
            if short_vol < MIN_SHORT_VOLUME {
                continue;
            }
            let long_vol = ctx.volume_in_window(token_id, LONG_WINDOW);
            let baseline_vol = long_vol - short_vol;
            if baseline_vol <= Decimal::ZERO {
                continue;
            }
            let baseline_secs = Decimal::from(LONG_WINDOW - SHORT_WINDOW);
            let baseline_rate = baseline_vol / baseline_secs;
            let short_rate = short_vol / Decimal::from(SHORT_WINDOW);
            if short_rate < baseline_rate * SPIKE_MULTIPLIER {
                continue;
            }

            let mid = (best_bid.price + best_ask.price) / dec!(2);
            let mut suggested_price = mid - dec!(0.005);
            if suggested_price <= dec!(0.01) {
                suggested_price = dec!(0.01);
            }

            let bucket = ctx.timestamp.timestamp() / 60;
            let short_prefix: String = token_id.chars().take(16).collect();
            let ratio = if baseline_rate.is_zero() {
                Decimal::ZERO
            } else {
                short_rate / baseline_rate
            };
            let reason = format!(
                "vol spike: short {} / baseline {} ({:.1}x)",
                short_vol, baseline_vol, ratio
            );
            let suggested = Signal::Sell {
                token_id: token_id.clone(),
                price: suggested_price,
                size: SUGGESTED_SIZE,
                urgency: Urgency::Medium,
            };
            signals.push(Signal::Alert {
                reason,
                suggested: Box::new(suggested),
                ttl_secs: ALERT_TTL_SECS,
                dedupe_key: format!("momentum_fade-{}-{}", short_prefix, bucket),
            });
        }
        if signals.is_empty() {
            return vec![Signal::Hold];
        }
        signals
    }

    fn on_fill(&mut self, _fill: &Fill) {}
}
